### PR TITLE
refactored OLAP

### PR DESCRIPTION
**Pull request by NLGithubWP**
1. Refactor the OLAP processing.
2. Remove some duplication .
3. Add more tests for OLAP query
4. Update the TimeDelimiter to "-" and min-max delimiter to " to "

### CODE REVIEW

**Comment by NLGithubWP** [comment]
This pr has some conflicts; I will fix them later. But the logic can be review firstly.

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelector.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+  private olapSelectionLayout selection;
+
+  private String timeRange;
+
+  // max and min query Time specified by user
+  private int maxQueryTime;
+  private int minQueryTime;
+
+  // time range split by system, used as key in map
```
Comment (hugy718) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelector.java:34]: These timeRanges are not used?

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    Map<AggregateType, AggregateFunc > aggMap = new HashMap<>();
+    for (AggregateType operator: aggregation.getOperators()) {
+      if (!checkOperatorIllegal(fieldType, operator)) {
+        throw new IllegalArgumentException(fieldName + " can not process " + operator);
+      }
+      AggregateFunc aggregator = AggregateFactory.generateAggregate(operator, fieldName);
+      aggMap.put(operator, aggregator);
+    }
+
+    // 3. init the result
```
Comment (hugy718) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java:283]: Here use RetUnit array directly, so CohortRet is not used?

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+      }
+      case NUMERIC: {
+        return String.valueOf(value);
+      }
+      default:
+        throw new UnsupportedOperationException();
+    }
+  }
+
+  /**
```
Comment (hugy718) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java:81]: Use a separate class for groupBy logic. Respect the Single responsibility principle gives us much cleaner codes. It is very hard to understand the code and where we can extend in the future for other types of processing now.

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/RangeFilter.java [hunk 4]
```diff
@@ -131,4 +132,14 @@ public void loadMetaInfo(MetaChunkRS metaChunkRS) {
     // for range Filter, no need to load info
   }
 
+  private static int convertPartToInt(String dataStr){
+    if (dataStr.contains("-")){
```
Comment (hugy718) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/RangeFilter.java:136]: ?? this is very adhoc, the logic should be put in a class to handle time range properly

**Review by hugy718 — COMMENTED** [review]
1. the iceberg tests are commented out, are they going to be rewritten?
2. ecommerce query result json file is empty, please remove it from the pr.

The logic seems to be not clean to me, likely because I never read the OLAP part before. In general, all logic is clumped together in the selector and aggregation. It is recommended to make it more modular, and abstract each standalone logic into a separate class, so that we can easily see what are operations currently support and how they can be extended in the future.

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelector.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    BitSet bs = bitSets.get(0);
+    for (int i = 1; i < bitSets.size(); i++) {
+      bs.or(bitSets.get(i));
+    }
+    return bs;
+  }
+
+  private void selectFields(BitSet bs, FieldRS field, Filter filter) {
+    // get local ids
+    InputVector fieldIn = field.getValueVector();
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelector.java:106]: Method `getValueVector`  is deprecated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+  private ChunkRS dataChunk;
+
+  private MetaChunkRS metaChunk;
+
+  private AggregateFactory aggregateFactory = new AggregateFactory();
+
+  private String getString(FieldRS field, MetaFieldRS metaField, int value, GroupType type) {
+    switch (type) {
+      case STRING: {
+        InputVector key = field.getKeyVector();
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java:70]: Method getKeyVector is deprecated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelectionLayout.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  @JsonProperty("type")
+  private FilterType dimension_type;
+
+  // values accepted in this field
+  @JsonProperty("acceptValue")
+  private List<String> values;
+
+  // sub-operation
+  @JsonProperty("fields")
+  private List<olapSelectionLayout> fields = new ArrayList<>();
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelectionLayout.java:58]: Circular reference ? This class is wired.

 According to the logic in olapSelectionLayout, it seems that there is a nested selectionLayout.
SelectionLayout contains fields which is a list of SelectionLayout.

According to the logic in olapSelector, the select method only support one-level process.

Maybe we should discuss about logic in this part and give a more clear implementation.

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+   * @param granularity when groupField is actionTime, group by granularity.
+   */
+  private void group(FieldRS field, BitSet bs, granularityType granularity) {
+    assert field.getFieldType() == FieldType.ActionTime;
+    long beg = System.currentTimeMillis();
+    // timeStr: bitMap, multiple rows may have same localID
+    Map<String, BitSet> timeStr2Bs = new HashedMap();
+
+    // get all local ids ( hash ), or value ( range )
+    InputVector values = field.getValueVector();
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java:138]: getValueVector is deprecated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+
+    // init projectedTuple.
+    ProjectedTuple tuple = new ProjectedTuple(projectedSchemaSet);
+
+    // 1. get the field type and data
+    String fieldName = aggregation.getFieldName();
+    FieldType fieldType = this.metaChunk.getMetaField(fieldName).getFieldType();
+
+    FieldRS field = this.dataChunk.getField(fieldName);
+    InputVector valueVec = field.getValueVector();
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java:271]: deprecated getValueVector()

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+   * @param metaField grouped filed read form one metaChunk
+   * @param type      group by type
+   */
+  private void group(FieldRS field, BitSet bs, MetaFieldRS metaField, GroupType type) {
+    long beg = System.currentTimeMillis();
+    Map<String, BitSet> group = new HashMap<>();
+    // localID: bitMap, multiple rows may have same localID
+    Map<Integer, BitSet> id2Bs = new HashedMap();
+    // get all local ids ( hash ), or value ( range )
+    InputVector values = field.getValueVector();
```
Comment (Zrealshadow) [inline cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java:95]: deprecated getValueVector

**Review by Zrealshadow — CHANGES_REQUESTED** [review]
The logic in OlapSelection is seems complex and hard to understand. Maybe you should give a proposal in issue like #50 . The after discussion, we can give a clear implementation.

Additionally, the separation of logic part and IO part is not clear. There are many redundant content between these two parts, For example in olapProcessor, there is still pointer which points to Layout class. These two parts are not fully decoupled.

Finally, some trivial mistake, for example, the name of class and file and some deprecated IPA

**Comment by NLGithubWP** [comment]
Re-design OLAP logic with renaming files and variables, modularize the functions, add comments, and fix format issues. The design pattern is shown as issue https://github.com/COOL-cohort/COOL/issues/122

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/OlapProcessor.java [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+   * timeRange, selection => groupBY => aggregate on each group
+   * @param cube the cube that stores the data we need
+   * @return the result of the query
+   */
+  public List<OlapRet> process(CubeRS cube) throws  Exception{
+
+    TableSchema tableschema = cube.getSchema();
+    // add this two schema into List
+
+    for (FieldSchema fieldSchema : tableschema.getFields()) {
```
[the remark this answers is not in the document]
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/OlapProcessor.java]: This is to reuse and to be compatible with the existing aggregators, which only accepts  `ProjectedTuple` and it requires  `projectedSchemaSet `.

File: datasets/olap-tpch/query.json [hunk 12]
```diff
@@ -1,17 +1,26 @@
 {
   "dataSource": "tpc-h-10g",
   "selection": {
-    "type": "and",
-    "dimension": null,
-    "values": null,
+    "filter_type": "and",
```
[the remark this answers is not in the document]
Comment (NLGithubWP) [reply datasets/olap-tpch/query.json]: Updated

File: cool-core/src/main/java/com/nus/cool/core/util/converter/DayIntConverter.java [hunk 13]
```diff
@@ -34,7 +34,7 @@ public class DayIntConverter implements NumericConverter {
 
   private DayIntConverter(){}
 
-  public static final DayIntConverter getInstance(){
+  public static DayIntConverter getInstance(){
```
[the remark this answers is not in the document]
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/util/converter/DayIntConverter.java]: Static methods may not requires final.

File: cool-core/src/main/java/com/nus/cool/core/io/readstore/ChunkRS.java [hunk 14]
```diff
@@ -144,6 +144,11 @@ public FieldRS getField(int i) {
     return this.fields[i];
   }
 
+  /**
+   * Get the filed information according to fild name
```
[the remark this answers is not in the document]
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/io/readstore/ChunkRS.java]: Updated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/storage/OlapRet.java [hunk 15]
```diff
@@ -0,0 +1,251 @@
+/*
```
[the remark this answers is not in the document]
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/storage/OlapRet.java]: Updated

[the remark this answers is not in the document]
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/storage/OlapRet.java]: OlapRet is actually the after-processing class. After OLAP processing, we call it to further process the result and show it to the user.  All integrating operators have nothing to do with OLAP logic.

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java (hunk 2 above)
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java]: OLAPRET is for after processing

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java (hunk 3 above)
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java]: Updated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/OlapQueryLayout.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.List;
+import lombok.Getter;
+import java.io.File;
+
+
+@Getter
+public class OlapQueryLayout {
+
+  // inner structure
+  public enum granularityType{
```
[the remark this answers is not in the document]
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/OlapQueryLayout.java]: Updated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelectionLayout.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+import com.nus.cool.core.cohort.refactor.filter.FilterLayout;
+import com.nus.cool.core.cohort.refactor.filter.FilterType;
+import com.nus.cool.core.io.readstore.MetaChunkRS;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import lombok.Data;
+
+@Data
+public class olapSelectionLayout {
```
[the remark this answers is not in the document]
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelectionLayout.java]: Updated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java (hunk 6 above)
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java]: Updated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelectionLayout.java (hunk 7 above)
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelectionLayout.java]: Updated in https://github.com/COOL-cohort/COOL/issues/122

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java (hunk 8 above)
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java]: Updated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java (hunk 10 above)
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java]: Updated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelector.java (hunk 5 above)
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapSelector.java]: Updated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/RangeFilter.java (hunk 4 above)
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/RangeFilter.java]: Abstract the TimeDelimiter into RangeFilter

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/RangeFilter.java [hunk 18]
```diff
@@ -87,7 +88,7 @@ public BitSet accept(String[] values) throws RuntimeException {
   @Override
   public boolean accept(Scope scope) throws RuntimeException {
     for (Scope u : acceptRangeList) {
-      if (u.isSubset(scope)) {
+      if (u.isSubset(scope) || u.isIntersection(scope)) {
```
[the remark this answers is not in the document]
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/filter/RangeFilter.java]: Updated

File: cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java (hunk 9 above)
Comment (NLGithubWP) [reply cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java]: Updated

**Pull request merged**